envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/ServiceTagFilter.kt

```diff
@@ -0,0 +1,129 @@
+    private val twoTagsCombinationsByService: Map<String, List<Pair<Regex, Regex>>>
+    private val threeTagsCombinationsByService: Map<String, List<Triple<Regex, Regex, Regex>>>
+
+    init {
+        properties.allowedTagsCombinations.forEach {
+            if (it.tags.size < 2 || it.tags.size > 3) {
+                throw IllegalArgumentException(
+                    "A tags combination must contain 2 or 3 tags. Combination with ${it.tags.size} tags found")
+            }
+        }
+        val combinationsByService = properties.allowedTagsCombinations
+            .groupBy({ it.serviceName }) { it.tags }
+
+        threeTagsCombinationsByService = combinationsByService
+            .mapValues { it.value.filter { it.size == 3 } }
+            .mapValues { it.value
+                .map { it.sorted().map { Regex(it) } }
+                .map { Triple(it[0], it[1], it[2]) }
+                .distinct()
+            }
+            .filterValues { it.isNotEmpty() }
+
+        twoTagsCombinationsByService = combinationsByService
+            .mapValues { it.value.filter { it.size == 2 } }
+            .mapValues { it.value
+                .map { it.sorted().map { Regex(it) } }
+                .map { it[0] to it[1] }
+                .distinct()
+            }
+            .mapValues { (serviceName, combinations) ->
+                combinations + threeTagsCombinationsByService[serviceName].orEmpty()
+                    .flatMap { it.getAllPairs() }
+            }
+            .filterValues { it.isNotEmpty() }
+    }
+
+    /**
+     * Transforms raw instance's tags to tags that should actually be add to instance metadata.
+     *  - discards tags that are not suitable for routing
+     *  - generates tags combinations that may be used for routing
+     *
+     * @return sequence of tags that should be add to instance's metadata or null if tag entry should not be add to
+     * instance's metadata.
+     */
+    fun getAllTagsForRouting(serviceName: String, instanceTags: Set<String>): Sequence<String>? {
+        val tags = filterTagsForRouting(instanceTags)
+        if (tags.isEmpty()) {
+            return null
+        }
+
+        val addPairs = isAllowedToMatchOnTwoTags(serviceName)
+        val addTriples = isAllowedToMatchOnThreeTags(serviceName)
+        val generatePairs = addPairs || addTriples
+
+        val tagsPairs = if (generatePairs) {
+            tags
+                .flatMap { tag1 -> tags
+                    .filter { it > tag1 }
+                    .filter { tag2 -> canBeCombined(serviceName, tag1, tag2) }
+                    .map { tag2 -> tag1 to tag2 }
+                }
+        } else {
+            emptyList()
+        }
+
+        val tagsPairsJoined = if (addPairs) {
+            tagsPairs.map { "${it.first},${it.second}" }.asSequence()
+        } else {
+            emptySequence()
+        }
+        val tagsTriplesJoined = if (addTriples) {
```

Comment:
I would extract `tagsPairsJoined` and `tagsTriplesJoined` code to method.

Reply:
done